internal/pkg/db/mongo/metadata.go

```diff
@@ -294,106 +297,103 @@ func (m MongoClient) UpdateDeviceReport(dr contract.DeviceReport) error {
 	s := m.session.Copy()
 	defer s.Close()
 
-	model := &models.DeviceReport{}
-	if err := model.FromContract(dr); err != nil {
+	var err error
+	var mapped models.DeviceReport
+	if err = mapped.FromContract(dr); err != nil {
 		return errors.New("FromContract failed")
 	}
 
-	var err error
-	if model.Id.Valid() {
-		err = s.DB(m.database.Name).C(db.DeviceReport).UpdateId(dr.Id, dr)
-	} else {
-		err = s.DB(m.database.Name).C(db.DeviceReport).Update(bson.M{"uuid": model.Uuid}, model)
+	var query bson.M
+	if query, err = idToBsonM(dr.Id); err != nil {
+		return err
 	}
-	return errorMap(err)
+	return errorMap(s.DB(m.database.Name).C(db.Addressable).Update(query, mapped))
 }
 
 func (m MongoClient) DeleteDeviceReportById(id string) error {
 	return m.deleteById(db.DeviceReport, id)
 }
 
 /* ----------------------------- Device ---------------------------------- */
-func (m MongoClient) AddDevice(d *contract.Device) error {
+func (m MongoClient) AddDevice(d contract.Device) (string, error) {
 	s := m.session.Copy()
 	defer s.Close()
+
 	col := s.DB(m.database.Name).C(db.Device)
 
 	// Check if the name exist (Device names must be unique)
 	count, err := col.Find(bson.M{"name": d.Name}).Count()
 	if err != nil {
-		return err
+		return "", err
 	}
 	if count > 0 {
-		return db.ErrNotUnique
+		return "", db.ErrNotUnique
 	}
-	ts := db.MakeTimestamp()
-	d.Created = ts
-	d.Modified = ts
-	d.Id = bson.NewObjectId()
 
-	addr, err := m.getAddressableByName(d.Addressable.Name)
-	if err != nil {
-		return err
+	var device models.Device
+	if err := device.FromContract(d, m, m, m, m); err != nil {
```

Comment:
The use of `m` here is kind of unfortunate. In all of these transform operations, you're passing in the MongoClient instance anyway. Could you remove the transform params and have the following method:

`func (d *Device) FromContract(from contract.Device, mc MongoClient)`

and then cast the `mc` to the appropriate transform where necessary within the function? For example:
```
aTransform := mc.(addressableTransform)
if d.Addressable, err = aTransform.AddressableToDBRef(aModel); err != nil {
    return
}
```

Reply:
Why is the existing implementation lacking? 

Design decision to separate transform contracts by type was based on interface segregation principle.  MongoClient happens to implement them all.  It doesn't have to in the future (if, for instance, we extract type-specific repositories from the current mono-repository that is MonoClient).
 
Passing MongoClient to disambiguate to individual interfaces in individual FromContract methods violates encapsulation -- it equates to passing around a global struct and letting callers pick from it what they want/need.

Reply:
Valid point.

`It doesn't have to in the future (if, for instance, we extract type-specific repositories from the current mono-repository that is MonoClient).`

This is something that occurred to me last night. Maybe it's not optimal to have the MongoClient implement those interfaces. Perhaps they should be their own types. On the other hand, Device is a special case thus far in the exercise and the only place where we've encountered this.

I won't block the PR as is, but I still think it's a code smell.